### PR TITLE
Release infra & doc changes for alpha and beta releases

### PR DESCRIPTION
**Pull request by ikehz**
@quinton-hoole @fgrzadkowski @zmerlynn

This is a proposal for changes I think need to be made to the release process to take care of #13339, #15483, and #13343.  Ignore everything below **Origin of the Sources**.  You can read it in Markdown [here](https://github.com/ihmccreery/kubernetes/blob/release-docs/docs/devel/releasing.md).  The two big things are:
1. We'll now have three separate processes, to be consistent with the changes in versioning we made in #15340:
   - Cutting an alpha release (`vX.Y.0-alpha.W`)
   - Cutting an official release (`vX.Y.Z`)
   - Branching a new release series (`vX.Y`)
2. Release infrastructure will live in `kubernetes/contrib`, not the main `kubernetes/kubernetes` repo.
   - I'm not wedded to this, but I do think it would be a good idea, and regardless, I think it's important that we separate our build infrastructure from our release infrastructure.

Please take a look and let me know what you think.

### CODE REVIEW

**Comment by ihmccreery** [comment]
I just realized that for any release in which we're doing commits, (right now, official releases,) we have to either:

1. freeze the branch we're taking from, so that we can do commits at HEAD of that branch,
2. create a new release-specific branch, (e.g. branch `v1.2.3`—we'd end up with ~35 branches by the end of 2016), or
3. not push the actual versioning commits, and just do that behind the scenes.

I think 1. seems the most reasonable, and it's what we currently do for releases.  But, if you were wondering, as I was, why we didn't do the whole versionizing-docs dance during alpha releases, this is why: we can't freeze master.

**Comment by ihmccreery** [comment]
@quinton-hoole @zmerlynn @fgrzadkowski

These changes are ready to be reviewed.  This automates and guarantees, (somewhat,) the processes necessary for alpha & beta releases (#13339 and #15483), plus changes in docs.

There's a bit more work to be done to make sure the build process works correctly, (`build/push-official-release.sh`,) but that will have to be cherrypicked into `release-1.0` and `release-1.1` (see #16529), so it needs to go into a separate PR.  This is only stuff that's version-agnostic.

Can I get a review?  Sorry it's so big.  There's a lot of Frankenstein-ed logic, but it had to be broken up into smaller bits for re-use, which is why it looks like a wholesale change.

**Comment by ihmccreery** [comment]
Take it back.  Per #16522 I need to update some stuff in here.

**Comment by bgrant0607** [comment]
Note that lots of doc changes are in progress for 1.1: #15893

**Comment by bgrant0607** [comment]
We need to be able to copy the docs from the release branch to gh-pages independently of (before, and maybe after) the release tag is created.

**Comment by ihmccreery** [comment]
@bgrant0607 04b6a51 has the docs changes we discussed this morning.

@quinton-hoole Do you have the capacity to review this?  I'd like to get it in so we can use it to cut `v1.1.1`.

**Comment by quinton-hoole** [comment]
@ihmccreery Sure, I can review today, before lunch.  Busy with another review right now.

**Comment by ihmccreery** [comment]
@k8s-bot test this please

**Comment by ihmccreery** [comment]
@k8s-bot unit test this please

File: docs/design/versioning.md [hunk 1]
```diff
@@ -35,23 +35,26 @@ Documentation for other releases can be found at
...
-* Kube 1.1.0: Final release, cut from the release-1.1.0 branch cut two weeks prior. Should occur between 3 and 4 months after 1.0.  1.1.1-beta will be tagged at the same time on the same branch.
+* Kube X.Y.0-alpha.W, W > 0: Alpha releases are released roughly every two weeks directly from the master branch. No cherrypick releases. If there is a critical bugfix, a new release from master can be created ahead of schedule.
+* Kube X.Y.Z-beta.W: When master is feature-complete for Kube X.Y, we will cut the release-X.Y branch 2 weeks prior to the desired X.Y.0 date and cherrypick only PRs essential to X.Y.  This cut will be marked as X.Y.0-beta.0, and master will be revved to X.Y+1.0-alpha.0.  If we're not satisfied with X.Y.0-beta.0, we'll release other beta releases, (X.Y.0-beta.W | W > 0) as necessary.
+* Kube X.Y.0: Final release, cut from the release-X.Y branch cut two weeks prior.  X.Y.1-beta.0 will be tagged at the same commit on the same branch.  X.Y.0 occur 3 to 4 months after X.Y-1.0.
```
Comment (quinton-hoole) [inline docs/design/versioning.md]: nit: The use of '-' t mean both 'minus' and as a prefix to 'alpha', 'beta' etc makes this a little confusing to read.  I don't have a good suggestion as to how better to express this, but perhaps you do?

Comment (ihmccreery) [inline docs/design/versioning.md]: Yeah, I thought about it for a while but couldn't think of anything better either, which is why I left it as is.

File: docs/design/versioning.md [hunk 2]
```diff
@@ -35,23 +35,26 @@ Documentation for other releases can be found at
...
+* Kube X.Y.Z, Z > 0: [Patch releases](#patches) are released as we cherrypick commits into the release-X.Y branch, (which is at X.Y.(Z-1)-beta.W,) as needed.  X.Y.Z is cut straight from the release-X.Y branch, and X.Y.Z+1-beta.0 is tagged on the same commit.
 
 ### Major version timeline
 
 There is no mandated timeline for major versions. They only occur when we need to start the clock on deprecating features. A given major version should be the latest major version for at least one year from its original release date.
 
+### CI version scheme
+
+* Continuous integration versions also exist, and are versioned off of alpha and beta releases.  X.Y.Z-alpha.W.C+aaaa is C commits after X.Y.Z-alpha.W, with an additional +aaaa build suffix added; X.Y.Z-beta.W.C+bbbb is C commits after X.Y.Z-beta.W, with an additional +bbbb build suffix added.  Furthermore, builds that are built off of a dirty build tree, (with things in the tree that are not checked it,) it will be appended with -dirty.
```
Comment (quinton-hoole) [inline docs/design/versioning.md]: Do we not want to avoid dirty builds entirely?

**Comment by ihmccreery** [comment]
@bgrant0607 See bc9feec for changes to allow multiple beta releases.

File: docs/devel/releasing.md [hunk 3]
```diff
@@ -42,31 +42,67 @@ after the first section.
...
+  * Official releases should be similar or identical to their respective beta
+    releases, so have a look at the cherry picks that have been merged since
+    the beta release and question everything you find.
+  * Official releases also require green tests, (see below).
+* New release series are also cut directly from `master`.
+  * **This is a big deal!**  If you're reading this doc for the first time, you
+    probably shouldn't be doing this release, and should talk to someone on the
+    release team.
+  * New release series cut a new release branch, `release-X.Y`, off of
+    `master`, and also release the first beta in the series, `vX.Y.0-beta`.
```
Comment (quinton-hoole) [inline docs/devel/releasing.md]: Don't you mean 'alpha' here?

File: docs/devel/releasing.md [hunk 4]
```diff
@@ -81,101 +117,106 @@ oncall.
...
+export VER="vX.Y.0-alpha.W"
 ```
 
-Where `v0.20.2-322-g974377b` is the git hash you decided on. This will become
-our (retroactive) branch point.
+then, run
 
-#### Branching, Tagging and Merging
+```console
+release/cut-official-release.sh "${VER}" "${GITHASH}"
```
Comment (quinton-hoole) [inline docs/devel/releasing.md]: Is this true?  Do we use cut-official-release.sh to cut alpha releases?  If so, how does it end up doing different things than are described below under 'Cutting and official release'?

File: release/build-official-release.sh [hunk 5]
```diff
@@ -88,29 +88,31 @@ ln -s ${KUBE_BUILD_DIR}/_output/release-tars/kubernetes.tar.gz ${KUBE_BUILD_DIR}
...
-echo "binary | hash alg | hash"
-echo "------ | -------- | ----"
-echo "\`kubernetes.tar.gz\` | md5 | \`${MD5}\`"
-echo "\`kubernetes.tar.gz\` | sha1 | \`${SHA1}\`"
-echo ""
-echo "     We'll fill in the release notes in the next stage."
-echo "  5) Ensure all the binaries are in place on GitHub and GCS before cleaning."
-echo "  6) (cd ${KUBE_BUILD_DIR}; make clean; cd -; rm -rf ${KUBE_BUILD_DIR})"
-echo ""
+cat <<- EOM
```
Comment (quinton-hoole) [inline release/build-official-release.sh]: This whole hunk is identical other than the change from using echo to using <<-, right?  Nothing real to review here?

File: release/cut-official-release.sh [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+# Sets DIR, INSTRUCTIONS
+function main() {
+  # Parse arguments
+  if [[ "$#" -ne 2 ]]; then
+    usage
+    exit 1
+  fi
+  local -r new_version=${1-}
+  # TODO(ihmccreery) Stop calling it githash; it's not a githash.
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: Can we fix this before merging?

File: release/cut-official-release.sh [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+  local -r githash=${2-}
+  DRY_RUN=true
+  if [[ "${3-}" == "--no-dry-run" ]]; then
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: This is not covered in in your instructions above.

File: release/cut-official-release.sh [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    local -r release_type='beta'
+    local -r version_major="${BASH_REMATCH[1]}"
+    local -r version_minor="${BASH_REMATCH[2]}"
+    local -r version_patch="${BASH_REMATCH[3]}"
+    local -r version_beta_rev="${BASH_REMATCH[4]}"
+  elif [[ "${new_version}" =~ $official_version_regex ]]; then
+    local -r release_type='official'
+    local -r version_major="${BASH_REMATCH[1]}"
+    local -r version_minor="${BASH_REMATCH[2]}"
+    local -r version_patch="${BASH_REMATCH[3]}"
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: nit: Can you perhaps remove some of the cut 'n paste code for release_type, version_major, version_minor, version_alpha_rev here?  For example by lifting common code out to above the 'if' statement?  Same comment applies to the regex's above.  They are all almost identical.  Can you factor out the common stuff?

File: release/cut-official-release.sh [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+  # Set the default umask for the release. This ensures consistency
+  # across our release builds.
+  #
+  # TODO(ihmccreery): This should be in our build process, not our release
+  # process.
+  local -r release_umask=${release_umask:-022}
+  umask "${release_umask}"
+
+  # Start a tmp file that will hold instructions for the user.
+  declare -r INSTRUCTIONS="/tmp/kubernetes-${release_type}-release-${new_version}-$(date +%s)-instructions"
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: It's bad karma to write directly to /tmp.  
Use mktmp?  Something like this: 
http://unix.stackexchange.com/questions/181937/how-create-a-temporary-file-in-shell-script

File: release/cut-official-release.sh [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+  if $DRY_RUN; then
+    cat > "${INSTRUCTIONS}" <<- EOM
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: This seems like gratuitous use of this syntax.  Would two echo's not be clearer?

File: release/cut-official-release.sh [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+Success!  You would now do the following, if not a dry run:
+
+EOM
+  else
+    cat > "${INSTRUCTIONS}" <<- EOM
+Success!  You must now do the following:
+
+EOM
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: Same comment as above?

File: release/cut-official-release.sh [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+  fi
+
+  local -r github="https://github.com/kubernetes/kubernetes.git"
+  declare -r DIR="/tmp/kubernetes-${release_type}-release-${new_version}-$(date +%s)"
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: Same comment as above.  Use mktmp?

File: release/cut-official-release.sh [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+    beta-release "${beta_version}"
+    git-push ${release_branch}
+  fi
+
+  echo
+  cat "${INSTRUCTIONS}"
+}
+
+function usage() {
+  echo "Usage: ${0} <version> <githash>"
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: What about --no-dry-run ?

File: release/cut-official-release.sh [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+  echo
+  echo "See docs/devel/releasing.md for more info."
+}
+
+function check-prereqs() {
+  SED=sed
+  if which gsed &>/dev/null; then
+    SED=gsed
+  fi
+  if ! ($SED --version 2>&1 | grep -q GNU); then
```
Comment (quinton-hoole) [inline release/cut-official-release.sh]: This all seems a little over-complicated?  Can you not just branch on the 'if which' statement above?

File: docs/devel/releasing.md (hunk 4 above)
Comment (quinton-hoole) [inline docs/devel/releasing.md]: Never mind.  Reading through the script, this now makes sense.

**Comment by quinton-hoole** [comment]
@ihmccreery Thanks for all this grungy work.  Looks mostly good to me, barring my relatively minor nits.  My bash sense of style is not the greatest :-), so it would be good to have a more bash expertish person give the bash parts a once-over for good style.  @ixdy @zmerlynn ?

File: docs/design/versioning.md (hunk 2 above)
Comment (ihmccreery) [inline docs/design/versioning.md]: I don't think so; we want devs to be able to build if they haven't checked everything into git, right?  Maybe I should make the heading "CI and dev version scheme"?  Is that what's confusing this?

File: docs/devel/releasing.md (hunk 3 above)
Comment (ihmccreery) [inline docs/devel/releasing.md]: Nope.  When we decide we want to branch v1.2, we'll immediately cut a v1.2.0-beta.0.  Though this should have gotten updated for versioned beta releases.  Done in a9fc43d.

File: release/build-official-release.sh (hunk 5 above)
Comment (ihmccreery) [inline release/build-official-release.sh]: Correct.

File: release/cut-official-release.sh (hunk 6 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: No, the change would have to be cherrypicked, so it would need to be a separate PR.

File: release/cut-official-release.sh (hunk 7 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: Fixed in 3466762.

File: release/cut-official-release.sh (hunk 8 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: While I agree with the intention, I think that would be a bad choice here, and would introduce more complexity than it would remove.  These regex's are different enough that it'd be pretty confusing.  For example, `BASH_REMATCH[3]` of the alpha regex is the alpha rev, whereas for the beta and official versions, it's the patch version, and for the series release, it doesn't exist.  I could just factor out major and minor versions, but I still think that would decrease readability and increase complexity.

File: release/cut-official-release.sh (hunk 10 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: This is now 7 lines instead of 2, so I think not, but if you feel strongly I'll change it.

File: release/cut-official-release.sh (hunk 11 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: I'd prefer consistency to doing exactly the right thing, but will change if you feel strongly.

File: release/cut-official-release.sh (hunk 14 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: This is copy-pasted from the old script.  Looking at it, I think this is the only way to take into account the possibility that GNU sed is masquerading as `sed`, (and not installed as gsed,) and the possibility that both `sed` and `gsed` are installed as themselves.

File: docs/design/versioning.md (hunk 2 above)
Comment (ihmccreery) [inline docs/design/versioning.md]: Fixed in ed7b42c.

File: release/cut-official-release.sh (hunk 13 above)
Comment (ihmccreery) [inline release/cut-official-release.sh]: Fixed in ed7b42c.

**Comment by ihmccreery** [comment]
@quinton-hoole PTAL.

**Comment by ihmccreery** [comment]
@quinton-hoole @zmerlynn doesn't have time to review this.

**Comment by quinton-hoole** [comment]
LGTM

**Comment by ihmccreery** [comment]
Removing lgtm label until we're through today and the kinks are worked out.

**Comment by mikedanese** [comment]
@eparis @lavalamp k8s-merge-bot is an unstoppable force

**Comment by ixdy** [comment]
@mikedanese makes me wonder: if @ihmccreery had pushed more commits after mergebot asked for the final e2e run, would it have noticed?

**Comment by ihmccreery** [comment]
This is a known issue: kubernetes/contrib#129.
